mirrord/agent/iptables/src/lib.rs

```diff
@@ -346,33 +368,94 @@ pub fn get_iptables(nftables: Option<bool>, ip6: bool) -> IPTablesWrapper {
-        );
-    });
+    let legacy = get_iptables(Some(false), ip6);
+    let nft = get_iptables(Some(true), ip6);
+
+    try_drop_cap_sys_module();
+
+    for (backend, is_nft) in [(&legacy, false), (&nft, true)] {
+        match MeshVendor::detect(backend) {
+            Ok(Some(mesh)) => {
+                tracing::info!(
+                    %mesh,
+                    command = backend.tables.cmd,
+                    "Detected mesh rules with one of the iptables backends. \
+                    Using this backend."
+                );
+                let _ = detected_nftables.set(is_nft);
+                return backend.clone();
+            }
+            Ok(None) => {
+                tracing::debug!(
+                    command = backend.tables.cmd,
+                    "No mesh rules detected with one of the iptables backends."
+                );
+            }
+            Err(error) => {
+                tracing::debug!(
+                    error = %error,
+                    command = backend.tables.cmd,
+                    "Failed to detect mesh rules with one of the iptables backends."
+                );
+            }
+        }
+    }
 
-    let nft_rules_present = nft_mesh.as_ref().is_ok_and(Option::is_some);
-    let _ = detected_nftables.set(nft_rules_present);
-    let wrapper = if nft_rules_present {
-        nft_wrapper
+    let legacy_works = legacy
+        .tables
+        .append("nat", "INPUT", "-p 255 -j RETURN")
+        .inspect_err(|error| {
+            tracing::debug!(
+                %error,
+                command = legacy.tables.cmd,
+                "Failed to add a dummy rule using iptables-legacy binary, \
+                assuming no kernel support for it.",
+            )
+        })
+        .is_ok();
+    let _ = detected_nftables.set(legacy_works.not());
+    let wrapper = if legacy_works {
+        let _ = legacy.tables.delete("nat", "INPUT", "-p 255 -j RETURN")
+            .inspect_err(|error| {
+                tracing::error!(
+                    %error,
+                    command = legacy.tables.cmd,
+                    "Failed to delete a dummy rule added when checking kernel support for iptables-legacy. \
+                    The rule should not affect any traffic."
+                )
+            });
+        legacy
     } else {
-        get_iptables(Some(false), ip6)
+        nft
     };
 
     tracing::info!(
-        cmd = wrapper.tables.cmd,
-        detect_mesh_in_nft_result = ?nft_mesh,
-        "Using iptables backend picked based on mesh rules detection."
+        command = wrapper.tables.cmd,
+        "Using iptables backend picked based on kernel support for iptables-legacy.",
     );
 
     wrapper
 }
 
+/// Drops [`Capability::CAP_SYS_MODULE`] from the current thread.
+///
+/// This will prevent the thread from loading kernel modules.
+fn try_drop_cap_sys_module() {
+    let has_cap = caps::has_cap(None, CapSet::Effective, Capability::CAP_SYS_MODULE)
```

Comment:
I also noticed that the cap check is thread specific, the agent can be multithreaded no? so do we care only when we are setting up iptables?